### ISSUE TITLE
Matching Service: Partitioned TaskList for higher throughput

### ISSUE THREAD

**Issue by samarabbas** [uber/cadence#52]

**Comment by rpasea** [comment]
Guys, what's the status of this feature?

From what I understand, currently a single matching service node distributes all the work belonging to a TaskList. What's the impact of this? Does anybody have some throughput measurements? How many tasks per sec can a matcher distribute?

Are there any workarounds for this? A simple one I can think of is duplicating TaskLists into and then starting their workflows in a round robin fashion. Are there such setups supported directly by the cadence client library?

**Comment by samarabbas** [comment]
@rpasea This task is currently not funded on our side nor we have an ETA on when we can start working on it.  Fortunately there is a very straightforward workaround as you pointed out.  When starting workers specify multiple tasklist and start a worker for each of them.  For example:
```
{application}-tasklist-0
{application}-tasklist-1
{application}-tasklist-2
```
When starting workflow you can randomly pick a tasklist and pass it in as parameter to workflow execution.  Then use this tasklist for scheduling activities from that workflow execution.  We don't have a high level client abstraction which does this automatically but it should be pretty easy to built.
Answer to throughput measurements for a single tasklist depends on what kind of hardware you are using for the deployment.  Here are some numbers which see on our production cluster:
Number of tasks dispatch (with SyncMatch): 400/sec
Number of tasks dispatch (without SyncMatch): 100/sec

Specs:
MatchingHost Memory: 8GB
MatchingHost Cores: 4
Cassandra Memory: 32 GB
Cassandra Cores: 28

**Comment by rpasea** [comment]
Thanks for the thorough response, Samar!

**Comment by dmetzgar** [comment]
Implemented as part of #2229. Assigned to @yux0 for testing and rollout.

**Comment by wxing1292** [comment]
core functionality is implemented and rolled out
remaining issue is tracked in #2473 #2483
